service/src/main/java/bio/terra/externalcreds/dataAccess/FenceAccountKeyDAO.java

```diff
@@ -0,0 +1,81 @@
+package bio.terra.externalcreds.dataAccess;
+
+import bio.terra.externalcreds.generated.model.Provider;
+import bio.terra.externalcreds.models.FenceAccountKey;
+import io.opentelemetry.instrumentation.annotations.WithSpan;
+import java.util.*;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.dao.support.DataAccessUtils;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.stereotype.Repository;
+
+@Repository
+@Slf4j
+public class FenceAccountKeyDAO {
+
+  private static final RowMapper<FenceAccountKey> FENCE_ACCOUNT_KEY_ROW_MAPPER =
+      ((rs, rowNum) ->
+          new FenceAccountKey.Builder()
+              .id(rs.getInt("id"))
+              .linkedAccountId(rs.getInt("linked_account_id"))
+              .keyJson(rs.getString("key_json"))
+              .expiresAt(rs.getTimestamp("expires_at"))
+              .build());
+
+  final NamedParameterJdbcTemplate jdbcTemplate;
+
+  public FenceAccountKeyDAO(NamedParameterJdbcTemplate jdbcTemplate) {
+    this.jdbcTemplate = jdbcTemplate;
+  }
+
+  @WithSpan
+  public Optional<FenceAccountKey> getFenceAccountKey(String userId, Provider provider) {
+    var namedParameters =
+        new MapSqlParameterSource()
+            .addValue("userId", userId)
+            .addValue("provider", provider.name());
+    var query =
+        "SELECT fence.id, fence.linked_account_id, fence.key_json, fence.expires_at FROM fence_account_key fence"
+            + " INNER JOIN linked_account la ON la.id = fence.linked_account_id"
+            + " WHERE la.user_id = :userId"
+            + " AND la.provider = :provider::provider_enum";
+    return Optional.ofNullable(
+        DataAccessUtils.singleResult(
+            jdbcTemplate.query(query, namedParameters, FENCE_ACCOUNT_KEY_ROW_MAPPER)));
+  }
+
+  @WithSpan
+  public FenceAccountKey insertFenceAccountKey(FenceAccountKey fenceAccountKey) {
```

Comment:
Can we rename this `upsertFenceAccountKey` and do the same `ON CONFLICT` logic?

Reply:
Is it just a conflict on the `linked_account_id`?

Reply:
Ah I had to add a unique constraint on the `linked_account_id` in order to use is as the conflict criteria